### DISCUSSION TITLE
Suspend workflow for long time period waiting for manual user input?

### DISCUSSION THREAD

**Discussion by justinmchase** [argoproj/argo-workflows#5754] [Q&A] [answered]
For example suppose I need a particular users Approval before I can continue.

My step is running and I update the database with an entry containing the workflowid and other metadata. This pending approval now shows up on a website for a particular user, they review it and click a button. The database is updated to reflect this action.

At that point I would like to resume the workflow. It would pull from the db and see that the approval was acquired and continue on with its work.

At a high level I'd like to know the following:

1. Can a workflow step cause the workflow to enter the suspended state rather than just signalign success or failure? Or in other words can you suspend the workflow from within the workflow step code?
2. Can you resume a workflow such that it would re-run the step that suspended the workflow?
3. How long can a workflow be suspended? Theoretically I would like my manual steps to cause a workflow to be suspended for up to a year

I would love to hear if argo has these capabilities, or if there are alternate ways to accomplish the desired results. Any help would be much appreciated, thanks!

**Comment by terrytangyuan** [comment, marked as answer]
Yes, check out how you can suspend a workflow here: https://github.com/argoproj/argo-workflows/tree/master/examples#suspending

**Reply by justinmchase** [reply]
I see. Thank you for the link!

 It sounds like argo supports a Step which itself is of type Suspend. So it would seem I need to flow the workflow into that step to suspend the workflow.

 To answer my questions more specifically I believe what you're saying is that:

1. **Yes** but not in the container code, rather it would have an _output parameter_ which would then be used by the condition of following steps to flow into a Suspend step.
2. **Yes**.  You can call an API to resume the workflow and it would flow into the next step in the pipeline.
3. It sounds like the answer is **indefinitely** but that the Suspend step has additional options as well.


I just wanted to paste my visualization of what we're talking about:

![image](https://user-images.githubusercontent.com/10974/116152727-e1a16000-a6ab-11eb-97d9-a7f56973bb85.png)

**Reply by terrytangyuan** [reply]
Yes they look correct to me. 👍🏻
